// secret-polar/dist/internal/core/project-structure.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.assertDir = exports.assertAllDirs = exports.getUserConfigPath = exports.isCwdProjectDir = exports.isCwdInsideProject = exports.multiImageVersion = exports.singleImageVersion = exports.TESTS_DIR = exports.SCRIPTS_DIR = exports.TS_SCHEMA_DIR = exports.SCHEMA_DIR = exports.TARGET_DIR = exports.CONTRACTS_OUT_DIR = exports.CACHE_DIR = exports.ARTIFACTS_DIR = exports.CONTRACTS_DIR = exports.JS_CONFIG_FILENAME = void 0;
const findUp = __importStar(require("find-up"));
const fs = __importStar(require("fs"));
const path_1 = require("path");
const fsp = fs.promises;
exports.JS_CONFIG_FILENAME = "polar.config.js";
exports.CONTRACTS_DIR = "contracts";
exports.ARTIFACTS_DIR = "artifacts";
exports.CACHE_DIR = (0, path_1.join)(exports.ARTIFACTS_DIR, ".cache");
exports.CONTRACTS_OUT_DIR = (0, path_1.join)(exports.ARTIFACTS_DIR, "contracts");
exports.TARGET_DIR = "target/wasm32-unknown-unknown/release/";
exports.SCHEMA_DIR = "artifacts/schema";
exports.TS_SCHEMA_DIR = "artifacts/typescript_schema";
exports.SCRIPTS_DIR = "scripts";
exports.TESTS_DIR = "test";
exports.singleImageVersion = "1.0.5";
exports.multiImageVersion = "1.0.5";
function isCwdInsideProject() {
    return Boolean(findUp.sync(exports.JS_CONFIG_FILENAME));
}
exports.isCwdInsideProject = isCwdInsideProject;
function isCwdProjectDir() {
    return Boolean(fs.existsSync(exports.JS_CONFIG_FILENAME));
}
exports.isCwdProjectDir = isCwdProjectDir;
function getUserConfigPath() {
    return findUp.sync(exports.JS_CONFIG_FILENAME) ?? undefined;
}
exports.getUserConfigPath = getUserConfigPath;
async function assertAllDirs() {
    const tasks = [];
    for (const d of [exports.CONTRACTS_DIR]) {
        tasks.push(assertDir(d));
    }
    await Promise.all(tasks);
}
exports.assertAllDirs = assertAllDirs;
async function assertDir(dirname) {
    try {
        await fsp.access(dirname, fs.constants.F_OK);
    }
    catch (e) {
        fs.mkdirSync(dirname, { recursive: true });
    }
}
exports.assertDir = assertDir;
