// secret-polar/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.polarTypes = exports.getLogs = exports.polarChai = exports.getAccountByName = exports.createAccounts = exports.Contract = void 0;
const account_1 = require("./lib/account");
Object.defineProperty(exports, "getAccountByName", { enumerable: true, get: function () { return account_1.getAccountByName; } });
const chai_1 = require("./lib/chai/chai");
Object.defineProperty(exports, "polarChai", { enumerable: true, get: function () { return chai_1.polarChai; } });
const createAccounts_1 = require("./lib/createAccounts");
Object.defineProperty(exports, "createAccounts", { enumerable: true, get: function () { return createAccounts_1.createAccounts; } });
const contract_1 = require("./lib/deploy/contract");
Object.defineProperty(exports, "Contract", { enumerable: true, get: function () { return contract_1.Contract; } });
const response_1 = require("./lib/response");
Object.defineProperty(exports, "getLogs", { enumerable: true, get: function () { return response_1.getLogs; } });
const polarTypes = __importStar(require("./types"));
exports.polarTypes = polarTypes;
